terraform destroy eks fails with Status [403 Forbidden]

when we delete an EKS cluster, the cluster actually gets deleted in the Rancher UI, but terraform does not know that the cluster got deleted in Rancher and it still tries to destroy every 10s. Rancher responds with a "403 Forbidden" error

```
module.k8s-rancher-sz-1.rancher2_cluster.eks_cluster: Destroying... (ID: c-qphhb)
module.k8s-rancher-sz-1.rancher2_cluster.eks_cluster: Still destroying... (ID: c-qphhb, 10s elapsed)
module.k8s-rancher-sz-1.rancher2_cluster.eks_cluster: Still destroying... (ID: c-qphhb, 20s elapsed)
module.k8s-rancher-sz-1.rancher2_cluster.eks_cluster: Still destroying... (ID: c-qphhb, 30s elapsed)
Releasing state lock. This may take a few moments...

Error: Error applying plan:

1 error(s) occurred:

* module.k8s-rancher-sz-1.rancher2_cluster.eks_cluster (destroy): 1 error(s) occurred:

* rancher2_cluster.eks_cluster: [ERROR] waiting for cluster (c-qphhb) to be removed: Bad response statusCode [403]. Status [403 Forbidden]. Body: [baseType=error, code=Forbidden, message=clusters.management.cattle.io "c-qphhb" is forbidden: User "u-67xj8" cannot get resource "clusters" in API group "management.cattle.io" at the cluster scope] from [https://rancher.infra.xxxxx.xxxxxxxx/v3/clusters/c-qphhb]
```

--- a/rancher2/resource_rancher2_cluster.go
+++ b/rancher2/resource_rancher2_cluster.go
@@ -3,10 +3,12 @@ package rancher2
 import (
 	"fmt"
 	"log"
+	"net/http"
 	"time"
 
 	"github.com/hashicorp/terraform/helper/resource"
 	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/rancher/norman/clientbase"
 	norman "github.com/rancher/norman/types"
 	managementClient "github.com/rancher/types/client/management/v3"
 )
@@ -254,6 +256,9 @@ func clusterStateRefreshFunc(client *managementClient.Client, clusterID string) 
 		err := client.APIBaseClient.ByID(managementClient.ClusterType, clusterID, obj)
 		if err != nil {
 			if IsNotFound(err) {
+				return obj, "removed", nil
+			} else if e, ok := err.(*clientbase.APIError); ok && e.StatusCode == http.StatusForbidden {
+				log.Printf("[INFO] Got access denied to the cluster, which means it has been removed")
 				return obj, "removed", nil
 			}
 			return nil, "", err

--- a/rancher2/resource_rancher2_cluster.go
+++ b/rancher2/resource_rancher2_cluster.go
@@ -3,12 +3,10 @@ package rancher2
 import (
 	"fmt"
 	"log"
-	"net/http"
 	"time"
 
 	"github.com/hashicorp/terraform/helper/resource"
 	"github.com/hashicorp/terraform/helper/schema"
-	"github.com/rancher/norman/clientbase"
 	norman "github.com/rancher/norman/types"
 	managementClient "github.com/rancher/types/client/management/v3"
 )
@@ -255,10 +253,7 @@ func clusterStateRefreshFunc(client *managementClient.Client, clusterID string) 
 		obj := &Cluster{}
 		err := client.APIBaseClient.ByID(managementClient.ClusterType, clusterID, obj)
 		if err != nil {
-			if IsNotFound(err) {
-				return obj, "removed", nil
-			} else if e, ok := err.(*clientbase.APIError); ok && e.StatusCode == http.StatusForbidden {
-				log.Printf("[INFO] Got access denied to the cluster, which means it has been removed")
+			if IsNotFound(err) || IsForbidden(err) {
 				return obj, "removed", nil
 			}
 			return nil, "", err

--- a/rancher2/resource_rancher2_cluster.go
+++ b/rancher2/resource_rancher2_cluster.go
@@ -253,6 +253,12 @@ func clusterStateRefreshFunc(client *managementClient.Client, clusterID string) 
 		obj := &Cluster{}
 		err := client.APIBaseClient.ByID(managementClient.ClusterType, clusterID, obj)
 		if err != nil {
+			// The IsForbidden check is used in the case the user performing the action does not have the
+			// right to retrieve the full list of clusters. If the user tries to retrieve the cluster that
+			// just got deleted, instead of getting a 404 not found response it will get a 403 forbidden
+			// eventhough it had the right to access the cluster before it was deleted. If we reach this
+			// code path, it means that the user had the right to access the cluster, delete it, hence
+			// meaning that the delete was successful.
 			if IsNotFound(err) || IsForbidden(err) {
 				return obj, "removed", nil
 			}

--- a/rancher2/util.go
+++ b/rancher2/util.go
@@ -169,6 +169,16 @@ func IsNotFound(err error) bool {
 	return clientbase.IsNotFound(err)
 }
 
+// IsForbidden checks if the given APIError is a Forbidden HTTP statuscode
+func IsForbidden(err error) bool {
+	apiError, ok := err.(*clientbase.APIError)
+	if !ok {
+		return false
+	}
+
+	return apiError.StatusCode == http.StatusForbidden
+}
+
 func splitTokenID(token string) string {
 	separator := ":"
 

